platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java

```diff
@@ -77,37 +56,12 @@ public R execute(UnitOfWork work, C context) {
         context.getBeforeExecutionState()
             .ifPresent(beforeExecutionState -> validateImplementations(work, beforeExecutionState, validationContext));
 
-        InternalProblems problemsService = validationContext.getProblemsService();
-        InternalProblemReporter reporter = problemsService.getInternalReporter();
-        List<InternalProblem> problems = validationContext.getProblems();
-
-        Map<Severity, ImmutableList<InternalProblem>> problemsMap = problems.stream()
-            .collect(
-                groupingBy(p -> p.getDefinition().getSeverity(),
-                    mapping(identity(), toImmutableList())));
-        List<InternalProblem> warnings = problemsMap.getOrDefault(WARNING, ImmutableList.of());
-        List<InternalProblem> errors = problemsMap.getOrDefault(ERROR, ImmutableList.of());
-
-        if (!warnings.isEmpty()) {
-            for (InternalProblem warning : warnings) {
-                reporter.report(warning);
-            }
-            warningReporter.recordValidationWarnings(work, warnings);
-        }
+        problemHandler.handleReportedProblems(work, validationContext);
 
-        if (!errors.isEmpty()) {
-            throwValidationException(work, validationContext, errors);
-        }
-
-        if (!warnings.isEmpty()) {
-            LOGGER.info("Invalidating VFS because {} failed validation", work.getDisplayName());
-            virtualFileSystem.invalidateAll();
-        }
```

Comment:
💭 This logic was moved to the `ExecutionProblemHandler` to be reused in `CheckDependenciesStep`.